src/analysis_and_optimization/Mem_pattern.ml

```diff
@@ -0,0 +1,476 @@
+  | Stmt.Fixed.Pattern.Decl
+      ({decl_id; decl_type= Type.Sized sized_type; _} as decl)
+    when Set.Poly.mem modifiable_set decl_id ->
+      Stmt.Fixed.Pattern.Decl
+        { decl with
+          decl_type=
+            Type.Sized (SizedType.modify_sizedtype_mem mem_pattern sized_type)
+        }
+  | NRFunApp (kind, (exprs : Typed.Meta.t Expr.Fixed.t list)) ->
+      let new_exprs = List.map ~f:mod_expr exprs in
+      let mod_kind = modify_kind mem_pattern modifiable_set kind new_exprs in
+      NRFunApp (mod_kind, new_exprs)
+  | Assignment
+      ( (name, ut, lhs)
+      , ( { pattern=
+              FunApp (CompilerInternal (FnReadParam (constrain_op, _)), args); _
+          } as assigner ) )
+    when Set.Poly.mem modifiable_set name ->
+      Assignment
+        ( (name, ut, List.map ~f:(Index.map mod_expr) lhs)
+        , { assigner with
+            pattern=
+              FunApp
+                ( CompilerInternal (FnReadParam (constrain_op, mem_pattern))
+                , List.map ~f:mod_expr args ) } )
+  | Assignment (((name : string), (ut : UnsizedType.t), lhs), rhs) ->
+      Assignment
+        ((name, ut, List.map ~f:(Index.map mod_expr) lhs), mod_expr rhs)
+  | IfElse (predicate, true_stmt, op_false_stmt) ->
+      IfElse
+        ( mod_expr predicate
+        , mod_stmt true_stmt
+        , Option.map ~f:mod_stmt op_false_stmt )
+  | Block stmts -> Block (List.map ~f:mod_stmt stmts)
+  | SList stmts -> SList (List.map ~f:mod_stmt stmts)
+  | For ({lower; upper; body; _} as loop) ->
+      Stmt.Fixed.Pattern.For
+        { loop with
+          lower= mod_expr lower; upper= mod_expr upper; body= mod_stmt body }
+  | TargetPE expr -> TargetPE (mod_expr expr)
+  | Return optional_expr -> Return (Option.map ~f:mod_expr optional_expr)
+  | Profile ((p_name : string), stmt) ->
+      Profile (p_name, List.map ~f:mod_stmt stmt)
+  | While (predicate, body) -> While (mod_expr predicate, mod_stmt body)
+  | Skip | Break | Continue | Decl _ -> pattern
+
+(**
+ * Modify statement patterns in the MIR from AoS <-> SoA and vice versa 
+ * @param mem_pattern A mem_pattern to modify expressions to. For the 
+ *  given memory pattern, this modifies 
+ *  statement patterns and expressions to it.
+ * @param stmt The statement to modify. 
+ * @param modifiable_set The name of the variable we are searching for.
+ *)
+and modify_stmt (mem_pattern : Common.Helpers.mem_pattern)
+    (Stmt.Fixed.({pattern; _}) as stmt) (modifiable_set : string Set.Poly.t) =
+  {stmt with pattern= modify_stmt_pattern mem_pattern pattern modifiable_set}
+
+(**
+ * Check one set against a base set to validate if anything in the set is 
+ * in the base set.
+ **)
+let check_names (base_set : string Set.Poly.t) (alt_set : string Set.Poly.t) :
+    bool =
+  let inter_set = Set.Poly.inter base_set alt_set in
+  let check_single_names (x : string) = Set.Poly.mem inter_set x in
+  Set.Poly.is_empty inter_set
+  || not (Set.Poly.for_all ~f:check_single_names alt_set)
```

Comment:
So IIUC this function returns true iff any element of alt_set is not in base_set, or if the alt_set is empty. I think you can do `Set.Poly.is_empty alt_set || not (Set.Poly.subset alt_set base_set)`
I think the comment is out of date

Reply:
Yep this works, updated doc as well!